File: ros2_ws/src/action_demo_py/action_demo_py/dishes_server.py
import time

import rclpy
from rclpy.action import ActionServer, CancelResponse, GoalResponse
from rclpy.node import Node

from action_demo_interfaces.action import DoDishes


class DoDishesServer(Node):
    def __init__(self) -> None:
        super().__init__("dishes_server")
        self._action_server = ActionServer(
            self,
            DoDishes,
            "dishes",
            execute_callback=self.execute_callback,
            goal_callback=self.goal_callback,
            cancel_callback=self.cancel_callback,
        )

    def goal_callback(self, goal_request: DoDishes.Goal) -> GoalResponse:
        self.get_logger().info(
            f"Received goal for dishwasher_id={goal_request.dishwasher_id}"
        )
        return GoalResponse.ACCEPT

    def cancel_callback(self, goal_handle) -> CancelResponse:
        self.get_logger().info("Received request to cancel goal")
        return CancelResponse.ACCEPT

    def execute_callback(self, goal_handle):
        feedback = DoDishes.Feedback()
        result = DoDishes.Result()

        for step in range(1, 6):
            if goal_handle.is_cancel_requested:
                result.total_dishes_cleaned = (step - 1) * goal_handle.request.dishwasher_id
                goal_handle.canceled()
                return result

            feedback.percent_complete = float(step * 20)
            goal_handle.publish_feedback(feedback)
            self.get_logger().info(f"Feedback {feedback.percent_complete:.1f}%")
            time.sleep(0.5)

        result.total_dishes_cleaned = goal_handle.request.dishwasher_id * 5
        goal_handle.succeed()
        self.get_logger().info(
            f"Goal succeeded, cleaned {result.total_dishes_cleaned} dishes"
        )
        return result


def main() -> None:
    rclpy.init()
    node = DoDishesServer()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()

File: ros2_ws/src/action_demo_cpp/src/dishes_server.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <thread>

#include "action_demo_interfaces/action/do_dishes.hpp"
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"

using namespace std::chrono_literals;

class DoDishesServer : public rclcpp::Node
{
public:
  using DoDishes = action_demo_interfaces::action::DoDishes;
  using GoalHandleDoDishes = rclcpp_action::ServerGoalHandle<DoDishes>;

  DoDishesServer()
  : Node("dishes_server")
  {
    action_server_ = rclcpp_action::create_server<DoDishes>(
      this,
      "dishes",
      std::bind(&DoDishesServer::handle_goal, this, std::placeholders::_1, std::placeholders::_2),
      std::bind(&DoDishesServer::handle_cancel, this, std::placeholders::_1),
      std::bind(&DoDishesServer::handle_accepted, this, std::placeholders::_1));
  }

private:
  rclcpp_action::GoalResponse handle_goal(
    const rclcpp_action::GoalUUID &,
    std::shared_ptr<const DoDishes::Goal> goal)
  {
    RCLCPP_INFO(get_logger(), "Received goal for dishwasher_id=%u", goal->dishwasher_id);
    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
  }

  rclcpp_action::CancelResponse handle_cancel(
    const std::shared_ptr<GoalHandleDoDishes> goal_handle)
  {
    (void)goal_handle;
    RCLCPP_INFO(get_logger(), "Received request to cancel goal");
    return rclcpp_action::CancelResponse::ACCEPT;
  }

  void handle_accepted(const std::shared_ptr<GoalHandleDoDishes> goal_handle)
  {
    std::thread(std::bind(&DoDishesServer::execute, this, std::placeholders::_1), goal_handle)
      .detach();
  }

  void execute(const std::shared_ptr<GoalHandleDoDishes> goal_handle)
  {
    const auto goal = goal_handle->get_goal();
    auto feedback = std::make_shared<DoDishes::Feedback>();
    auto result = std::make_shared<DoDishes::Result>();

    for (int step = 1; step <= 5; ++step) {
      if (goal_handle->is_canceling()) {
        result->total_dishes_cleaned = static_cast<uint32_t>((step - 1) * goal->dishwasher_id);
        goal_handle->canceled(result);
        RCLCPP_INFO(get_logger(), "Goal canceled");
        return;
      }

      feedback->percent_complete = static_cast<float>(step) * 20.0F;
      goal_handle->publish_feedback(feedback);
      RCLCPP_INFO(get_logger(), "Feedback %.1f%%", feedback->percent_complete);
      std::this_thread::sleep_for(500ms);
    }

    result->total_dishes_cleaned = goal->dishwasher_id * 5;
    goal_handle->succeed(result);
    RCLCPP_INFO(get_logger(), "Goal succeeded, cleaned %u dishes", result->total_dishes_cleaned);
  }

  rclcpp_action::Server<DoDishes>::SharedPtr action_server_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<DoDishesServer>());
  rclcpp::shutdown();
  return 0;
}
